compiler/rustc_typeck/src/check/method/mod.rs

```diff
@@ -198,6 +201,66 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let pick =
             self.lookup_probe(span, segment.ident, self_ty, call_expr, ProbeScope::TraitsInScope)?;
 
+        if span.edition() < Edition::Edition2021 {
+            if let sym::try_into = segment.ident.name {
+                if let probe::PickKind::TraitPick = pick.kind {
+                    if !matches!(self.tcx.crate_name(pick.item.def_id.krate), sym::std | sym::core)
+                    {
+                        self.tcx.struct_span_lint_hir(
+                            FUTURE_PRELUDE_COLLISION,
+                            call_expr.hir_id,
+                            call_expr.span,
+                            |lint| {
+                                let sp = call_expr.span;
+                                let trait_name =
+                                    self.tcx.def_path_str(pick.item.container.assert_trait());
+
+                                let mut lint = lint.build(&format!(
+                                    "trait method `{}` will become ambiguous in Rust 2021",
+                                    segment.ident.name
+                                ));
+
+                                if let Ok(self_expr) =
+                                    self.sess().source_map().span_to_snippet(self_expr.span)
+                                {
+                                    let derefs = "*".repeat(pick.autoderefs);
+                                    let self_adjusted = match pick.autoref_or_ptr_adjustment {
+                                        Some(probe::AutorefOrPtrAdjustment::Autoref {
+                                            mutbl: Mutability::Mut, ..
+                                        }) => format!("&mut {}{}", derefs, self_expr),
+                                        Some(probe::AutorefOrPtrAdjustment::Autoref {
+                                            mutbl: Mutability::Not, ..
+                                        }) => format!("&{}{}", derefs, self_expr),
+                                        Some(probe::AutorefOrPtrAdjustment::ToConstPtr) | None
```

Comment:
OK, so, `ToConstPtr` is used for when you have a `*mut T` value but the method is implemented for `*const T`. You could translate it to `str::ptr::addr_of!(*X)` (note the extra `*`).

Reply:
But we should try to make a test for it. This probably never happens in practice.

Reply:
My impression was `*mut T` -> `*const T` coercion of arguments would handle that, however I'll look into writing a test case for that

Reply:
I stand corrected, that is not the case, will fix that. Thanks!

Reply:
`str::ptr::addr_of!(*X)` doesn't seem to work. I was hesitant to try it due to being unsure if it follows the same semantics as `&*(&mut x)` (i.e. doesn't actually dereference).

According to the docs this is the intende behaviord

> Note, however, that the expr in addr_of!(expr) is still subject to all the usual rules. In particular, addr_of!(*ptr::null()) is Undefined Behavior because it dereferences a null pointer.

So I believe in this case, it will need to be a `*mut` -> `*const` as-cast?